subprojects/worker-processes/src/main/java/org/gradle/process/internal/worker/messaging/WorkerConfigSerializer.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright 2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.process.internal.worker.messaging;
+
+import org.gradle.api.Action;
+import org.gradle.api.logging.LogLevel;
+import org.gradle.internal.io.ClassLoaderObjectInputStream;
+import org.gradle.internal.remote.internal.inet.MultiChoiceAddress;
+import org.gradle.internal.remote.internal.inet.MultiChoiceAddressSerializer;
+import org.gradle.internal.serialize.Decoder;
+import org.gradle.internal.serialize.Encoder;
+import org.gradle.internal.serialize.InputStreamBackedDecoder;
+import org.gradle.internal.serialize.OutputStreamBackedEncoder;
+import org.gradle.internal.serialize.Serializer;
+import org.gradle.process.internal.worker.WorkerProcessContext;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+
+/**
+ * Serializes and de-serializes {@link WorkerConfig}s.
+ */
+public class WorkerConfigSerializer implements Serializer<WorkerConfig> {
+
+    public static final WorkerConfigSerializer INSTANCE = new WorkerConfigSerializer();
+    private WorkerConfigSerializer() {
+        // Private to enforce singleton.
+    }
+
+    public void writeTo(WorkerConfig config, DataOutputStream dos) throws IOException {
+        OutputStreamBackedEncoder encoder = new OutputStreamBackedEncoder(dos);
+        new WorkerConfigSerializer().write(encoder, config);
+        encoder.flush();
+    }
+
+    public WorkerConfig readFrom(DataInputStream dis) throws IOException {
+        return read(new InputStreamBackedDecoder(dis));
+    }
+
+    @Override
+    public WorkerConfig read(Decoder decoder) throws IOException {
+        LogLevel logLevel = LogLevel.values()[decoder.readSmallInt()];
+        boolean shouldPublishJvmMemoryInfo = decoder.readBoolean();
+        String gradleUserHomeDirPath = decoder.readString();
+        MultiChoiceAddress serverAddress = new MultiChoiceAddressSerializer().read(decoder);
```

Comment:
If keeping the `WorkerConfigSerializer` around as a Singleton, should probably keep this serializer around as a field.

Reply:
But also, what's the benefit to keeping this around as a Singleton vs. initializing a new object when needed?

Reply:
Really, there should be no instance at all and this should all be static code. But since I chose to implement `Serializer`, there must be an instance. In reality having a singleton instance is not saving much work at all so I suppose I could remove the singleton entirely and always instantiate a new serializer.

Reply:
I would remove the singleton and inline the writeTo/readFrom methods in the places they're used instead.  I believe that's usually how we use serializers

Reply:
Yeah I was thinking the `writeTo/readFrom` were a little extra here. Though I do like them I will remove them for the sake of consistency.